src/mlpack/methods/ann/rbm/spike_slab_rbm_policy_impl.hpp

```diff
@@ -0,0 +1,356 @@
+
+#ifndef MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+#define MLPACK_METHODS_ANN_RBM_SPIKE_SLAB_RBM_POLICY_IMPL_HPP
+
+#include "spike_slab_rbm_policy.hpp"
+
+namespace mlpack {
+namespace ann {
+template<typename InputDataType, typename OutputDataType>
+inline SpikeSlabRBMPolicy<InputDataType, OutputDataType>
+      ::SpikeSlabRBMPolicy(const size_t visibleSize,
+      const size_t hiddenSize,
+      const size_t poolSize,
+      const double slabPenalty, 
+      double radius):
+      visibleSize(visibleSize),
+      hiddenSize(hiddenSize),
+      poolSize(poolSize),
+      slabPenalty(slabPenalty),
+      radius(radius)
+{
+  parameter.set_size(visibleSize * hiddenSize * poolSize +
+      visibleSize + hiddenSize);
+
+  visibleMean.set_size(visibleSize, 1);
+  spikeMean.set_size(hiddenSize, 1);
+  spikeSamples.set_size(hiddenSize, 1);
+  slabMean.set_size(poolSize, hiddenSize);
+  invSlabPenalty = 1.0 / slabPenalty;
+};
+
+// Reset function
+template<typename InputDataType, typename OutputDataType>
+inline void SpikeSlabRBMPolicy<InputDataType, OutputDataType>::Reset()
+{
+  // Weight shape D * K * N
+  weight = arma::cube(parameter.memptr(), visibleSize, poolSize, hiddenSize,
+      false, false);
+
+  // spike bias shape N * 1
+  spikeBias = arma::mat(parameter.memptr() + weight.n_elem, hiddenSize, 1,
+      false, false);
+
+  // visible penalty 1 * 1 => D * D(when used)
+  visiblePenalty = arma::mat(parameter.memptr() + weight.n_elem +
+      spikeBias.n_elem, 1, 1, false, false);
+}
+
+/**
+ * Free energy of the spike and slab variable
+ * the free energy of the ssRBM is given my
+ * $v^t$$\Delta$v - $\sum_{i=1}^N$ 
+ * $\log{ \sqrt{\frac{(-2\pi)^K}{\prod_{m=1}^{K}(\alpha_i)_m}}}$ -
+ * $\sum_{i=1}^N \log(1+\exp( b_i +
+ * \sum_{m=1}^k \frac{(v(w_i)_m^t)^2}{2(\alpha_i)_m})$
+ *
+ * @param input the visible layer
+ */ 
+template<typename InputDataType, typename OutputDataType>
+inline double SpikeSlabRBMPolicy<InputDataType, OutputDataType>
+    ::FreeEnergy(InputDataType&& input)
+{
+  assert(input.n_rows == visibleSize);
+  assert(input.n_cols == 1);
+
+  scalarVisiblePenalty = visiblePenalty(0, 0);
```

Comment:
Yes you have to initialise it again and again since visible Penalty can change. Or i could just reference to visiblePenalty(0,0) variable with scalarVisiblePenalty as double pointer type.